### PR TITLE
opencascade: new variants

### PR DESCRIPTION
**Pull request by wdconinc**
OpenCascade has several major modules and not every application needs all of them. This adds variants for the various modules.

It also updates the 3rdparty dependency treatment.

### CODE REVIEW

**Comment by wdconinc** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/opencascade/package.py [hunk 1]
```diff
@@ -67,59 +74,134 @@ class Opencascade(CMakePackage):
         when="@:7.7.0",
     )
 
+    # Modules, per DAG at https://dev.opencascade.org/doc/refman/html/
+    variant("foundation_classes", default=True, description="Build Foundation Classes module")
+    variant(
+        "modeling_data",
+        default=True,
+        description="Build Modeling Data module",
+        when="foundation_classes",
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/opencascade/package.py]:
```suggestion
        when="+foundation_classes",
```

File: var/spack/repos/builtin/packages/opencascade/package.py [hunk 2]
```diff
@@ -67,59 +74,134 @@ class Opencascade(CMakePackage):
         when="@:7.7.0",
     )
 
+    # Modules, per DAG at https://dev.opencascade.org/doc/refman/html/
+    variant("foundation_classes", default=True, description="Build Foundation Classes module")
```
Comment (alecbcs) [inline var/spack/repos/builtin/packages/opencascade/package.py]: @wdconinc is there a use case where someone might not want to build the foundation classes?

Comment (wdconinc) [reply var/spack/repos/builtin/packages/opencascade/package.py]: Yeah, there isn't. I'll remove that variant.

**Comment by wdconinc** [comment]
@spackbot fix style

**Comment by wdconinc** [comment]
FYI, the splitting into different modules is a first step towards making some of the dependencies dependent on the enabled modules. E.g. `+vtk` doesn't make sense unless `+visualization`, but `+tbb` does make sense for the foundation classes. I can either do all that work in this PR or in a different one. LMK.

**Review by alecbcs — DISMISSED** [review]
Looks good to me. Thank you for the contribution @wdconinc! Would you like to continue work on this PR to segment out the dependencies or add it in a follow up?

File: var/spack/repos/builtin/packages/opencascade/package.py [hunk 3]
```diff
@@ -67,59 +74,129 @@ class Opencascade(CMakePackage):
...
-        if "+rapidjson" in self.spec:
-            args.append("-DUSE_RAPIDJSON=ON")
-            args.append("-D3RDPARTY_RAPIDJSON_DIR=%s" % self.spec["rapidjson"].prefix)
-        else:
-            args.append("-DUSE_RAPIDJSON=OFF")
+        # Disable documentation building
+        args.append("-DBUILD_DOC_Overview=OFF")
+
+        # Always build the foundation classes
+        args.append(self.define("BUILD_MODULE_foundation_classes", true))
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/opencascade/package.py]:
```suggestion
        args.append(self.define("BUILD_MODULE_FoundationClasses", true))
```

**Comment by wdconinc** [comment]
I just compared my notes and there was one issue in this PR that I just fixed, but the rest is for a future PR (i.e. adding conditionals to the e.g. `gl`  variants and dependencies). This is ready for merging.

**Review by alecbcs — APPROVED** [review]
Looks good to me!

**Pull request merged**